lib/spack/spack/spec.py

```diff
@@ -2958,16 +2958,15 @@ def do_parse(self):
                         # We're parsing an anonymous spec beginning with a
                         # key-value pair.
                         if not specs:
+                            self.push_tokens([self.previous, self.token])
+                            self.previous = None
                             specs.append(self.spec(None))
-                        self.expect(VAL)
-                        # Raise an error if the previous spec is already
-                        # concrete (assigned by hash)
-                        if specs[-1]._hash:
-                            raise RedundantSpecError(specs[-1],
-                                                     'key-value pair')
-                        specs[-1]._add_flag(
-                            self.previous.value, self.token.value)
-                        self.previous = None
+                        else:
+                            if specs[-1].concrete:
+                                # Trying to add k-v pair to spec from hash
+                                raise RedundantSpecError(specs[-1],
+                                                         'key-value pair')
+                            self.unexpected_token()
```

Comment:
@tgamblin I don't think it is possible to reach this line, I added it out of an abundance of caution. Do you want me to remove it?

I'm also not sure which gcc issue you're referring to. Is it from one of the PRs I called out here?